language_service/src/hover.rs

```diff
@@ -188,18 +193,38 @@ impl Visitor<'_> for HoverVisitor<'_> {
 }
 
 fn markdown_with_doc(doc: &Rc<str>, code: impl Display) -> String {
-    if doc.is_empty() {
+    let parsed_doc = parse_doc(doc);
+    if parsed_doc.summary.is_empty() {
         markdown_fenced_block(code)
     } else {
         format!(
             "{}
 {}",
-            doc,
+            parsed_doc.summary,
             markdown_fenced_block(code)
         )
     }
 }
 
+fn parse_doc(doc: &str) -> Documentation {
+    let summary_re =
+        Regex::new(r"(^|(\r?\n))\s*#\s*((S|s)ummary)[\s\r\n]*").expect("Invalid regex");
+    let header_re = Regex::new(r"\r?\n\s*#\s*(\w+)[\s\r\n]*").expect("Invalid regex");
+    let summary = match summary_re.find(doc) {
+        Some(summary_header) => {
+            let start = summary_header.end();
+            match header_re.find(&doc[start..]) {
+                Some(next_header) => &doc[start..(start + next_header.start())],
```

Comment:
This seems a little clunky. You could just use one regex and a capture group to pull out the summary details, rather than use two, start where one ends, run to the end if second doesn't match, etc..

Try the below on `doc`. It should give you one capture group which contains the Summary content that starts from after the `# Summary` line until the next header, or the end of the string if no next header.

`(?m)(?:^# Summary$)([\s\S]+?)(?:(^# .*)|\z)`

Reply:
Yes this was a bit clunky and I was looking for a better alternative. Using the regex flags to do multi-line really helps a lot and lets me use the regex capture. I've made some small changes to your proposed regex, but it is largely the same. Notably, I wanted the non-capture submatches to take away leading and trailing whitespace from the captured summary (this also prevents extra newlines from showing up in the capture), and I noticed that, while you do need at least one space after the `#`, markdown seems to support/ignore several spaces more as well, so I wanted to match at least one space after the `#`.
This is the regex I have currently: `(?mi)(?:^#\s+Summary\s*$\s*)([\s\S]*?)(?:\s*(^# .*)|\z)`

Reply:
I'd consider anything other than one space between "#" and "Summary" to be malformed, and that we shouldn't worry about it (markdown linting should catch that - same with trailing whitespace after "Summary").

It's weird to have a (non-greedy) match ending in optional whitespace, followed by another capture group that starts with optional whitespace. Just for the ease of reading the regular expression (as well as possible perf implications) it's good to keep them as simple as possible. Perhaps just the following minor tweaks `(?mi)(?:^# Summary$)([\s\S]+?)(?:(^# .*)|\z)` and then on the capture group just use `trim()` method on the str to remove leading and trailing whitespace? (Though looking back, I see that's just the same regex I proposed before 🤣 ).

Reply:
Okay the regex enforces exactly one space after `#` and uses `trim()` to remove leading/trailing whitespace in the summary.